### PR TITLE
7903448: JMH: Override the use of compiler hints using a system property

### PR DESCRIPTION
**Pull request by gilles-duboscq**
Using `-Djmh.compilerhints.mode=true` will force the use of compiler hints. `-Djmh.compilerhints.mode=false` will prevent the use of compiler hints. When the system property is not set, the current beahviour based on vm name and version will be used by default.

This is especially useful when working with uncommon / rare JVMs or when adding compiler hints support to a new JVM.

I considered auto-detecting support for compiler hints but that didn't fit very well in the current code and required much more intrusive changes.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [CODETOOLS-7903448](https://bugs.openjdk.org/browse/CODETOOLS-7903448): JMH: Override the use of compiler hints using a system property

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jmh.git pull/96/head:pull/96` \
`$ git checkout pull/96`

Update a local copy of the PR: \
`$ git checkout pull/96` \
`$ git pull https://git.openjdk.org/jmh.git pull/96/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 96`

View PR using the GUI difftool: \
`$ git pr show -t 96`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jmh/pull/96.diff">https://git.openjdk.org/jmh/pull/96.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jmh/pull/96#issuecomment-1458491507)

### CODE REVIEW

**Comment by gilles-duboscq** [comment]
Good idea. I'll do

**Comment by gilles-duboscq** [comment]
I have changed the `jmh.compilerhints.mode` system property to access the 3 suggested values and implemented it following the implementation of the blackhole modes.

File: jmh-core/src/main/java/org/openjdk/jmh/runner/CompilerHints.java [hunk 1]
```diff
@@ -188,10 +228,12 @@ public static List<String> getCompileCommandFiles(List<String> command){
...
     public static void addCompilerHints(List<String> command) {
-        if (!isHintCompatibleVM()) {
-            System.err.println("WARNING: Not a HotSpot compiler command compatible VM (\""
-                    + System.getProperty("java.vm.name") + "-" + System.getProperty("java.version")
-                    + "\"), compilerHints are disabled.");
+        if (!compilerHintsEnabled()) {
+            if (compilerHintsSelect() == CompilerHintsSelect.AUTO) {
+                System.err.println("WARNING: Not a HotSpot compiler command compatible VM (\""
+                        + System.getProperty("java.vm.name") + "-" + System.getProperty("java.version")
+                        + "\"), compilerHints are disabled.");
```
Comment (shipilev) [inline jmh-core/src/main/java/org/openjdk/jmh/runner/CompilerHints.java:235]: While you are at it, please change "compilerHints" -> "compiler hints".

**Review by shipilev — CHANGES_REQUESTED** [review]
I have more comments, apologies :)

**Comment by shipilev** [comment]
Ping @gilles-duboscq -- do you want to complete this, or want me to take over?

**Comment by shipilev** [comment]
> Other things have taken over but i was still planning to get back to this within the next 1-2 weeks. If somebody else needs that faster, feel free to take over. Thanks for the review feedback.

That's fine, there is no time pressure to complete this work. Looking forward to new version in a few weeks. I plan to have the next JMH release some time in May.

File: jmh-core/src/main/java/org/openjdk/jmh/runner/CompilerHints.java [hunk 2]
```diff
@@ -188,10 +228,12 @@ public static List<String> getCompileCommandFiles(List<String> command){
      * @param command all -XX:CompileCommandLine args will be removed and a merged file will be set
      */
     public static void addCompilerHints(List<String> command) {
-        if (!isHintCompatibleVM()) {
-            System.err.println("WARNING: Not a HotSpot compiler command compatible VM (\""
-                    + System.getProperty("java.vm.name") + "-" + System.getProperty("java.version")
-                    + "\"), compilerHints are disabled.");
+        if (!compilerHintsEnabled()) {
+            if (compilerHintsSelect() == CompilerHintsSelect.AUTO) {
```
[the remark this answers is not in the document]
Comment (gilles-duboscq) [reply jmh-core/src/main/java/org/openjdk/jmh/runner/CompilerHints.java:232]: I can reduce the depth by printing the message when `compilerHintsSelect() == CompilerHintsSelect.AUTO_OFF)` but i still need another `if` to return early in the `FORCE_OFF` case.

File: jmh-core/src/main/java/org/openjdk/jmh/runner/CompilerHints.java (hunk 1 above)
Comment (gilles-duboscq) [reply jmh-core/src/main/java/org/openjdk/jmh/runner/CompilerHints.java]: done

File: jmh-core/src/main/java/org/openjdk/jmh/runner/CompilerHints.java [hunk 3]
```diff
@@ -101,11 +102,50 @@ private CompilerHints(String file, String resource) {
         hints = Collections.unmodifiableSet(read());
     }
 
+    public enum CompilerHintsSelect {
+        FORCE_ON("Forced on"),
+        FORCE_OFF("Forced off"),
+        AUTO("Automatically selected");
```
[the remark this answers is not in the document]
Comment (gilles-duboscq) [reply jmh-core/src/main/java/org/openjdk/jmh/runner/CompilerHints.java]: done

File: jmh-core/src/main/java/org/openjdk/jmh/runner/CompilerHints.java [hunk 4]
```diff
@@ -390,7 +432,9 @@ private static void debug(String msg) {
         }
     }
 
-    public static void printBlackhole(PrintStream out) {
+    public static void printHints(PrintStream out) {
+        out.print("# Compiler hints: " + (compilerHintsEnabled() ? "enabled" : "disabled") + " (" + compilerHintsSelect().desc() + ")");
+        out.println();
```
[the remark this answers is not in the document]
Comment (gilles-duboscq) [reply jmh-core/src/main/java/org/openjdk/jmh/runner/CompilerHints.java]: done

**Review by shipilev — APPROVED** [review]
This version looks good!

There are some flukes in GHA testing, but they all seem to be environmental.

**Pull request closed without merging**